collect_app/src/test/java/org/odk/collect/android/widgets/utilities/ActivityGeoDataRequesterTest.java

```diff
@@ -0,0 +1,183 @@
+import static org.odk.collect.android.widgets.support.GeoWidgetHelpers.assertGeoPolyBundleArgumentEquals;
+import static org.odk.collect.android.widgets.support.GeoWidgetHelpers.getRandomDoubleArray;
+import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithReadOnly;
+import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.widgetTestActivity;
+import static org.odk.collect.android.widgets.utilities.ActivityGeoDataRequester.ACCURACY_THRESHOLD;
+import static org.robolectric.Shadows.shadowOf;
+
+@RunWith(RobolectricTestRunner.class)
+public class ActivityGeoDataRequesterTest {
+    private final FakePermissionUtils permissionUtils = new FakePermissionUtils();
+    private final ActivityGeoDataRequester activityGeoDataRequester = new ActivityGeoDataRequester(permissionUtils);
+    private final FakeWaitingForDataRegistry waitingForDataRegistry = new FakeWaitingForDataRegistry();
+    private final GeoPointData answer = new GeoPointData(getRandomDoubleArray());
+
+    private TestScreenContextActivity testActivity;
+    private ShadowActivity shadowActivity;
+    private FormEntryPrompt prompt;
+    private FormIndex formIndex;
+    private QuestionDef questionDef;
+
+    @Before
+    public void setUp() {
+        testActivity = widgetTestActivity();
+        shadowActivity = shadowOf(testActivity);
+
+        prompt = promptWithReadOnly();
+        formIndex = mock(FormIndex.class);
+        questionDef = mock(QuestionDef.class);
+
+        permissionUtils.setPermissionGranted(true);
+        when(prompt.getQuestion()).thenReturn(questionDef);
+        when(prompt.getIndex()).thenReturn(formIndex);
+    }
+
+    @Test
+    public void whenPermissionIsNotGranted_requestGeoPoint_doesNotLaunchAnyIntent() {
+        permissionUtils.setPermissionGranted(false);
+        activityGeoDataRequester.requestGeoPoint(testActivity, prompt, waitingForDataRegistry);
+
+        assertNull(shadowActivity.getNextStartedActivity());
+        assertTrue(waitingForDataRegistry.waiting.isEmpty());
+    }
+
+    @Test
+    public void whenPermissionIsNotGranted_requestGeoShape_doesNotLaunchAnyIntent() {
+        permissionUtils.setPermissionGranted(false);
+        activityGeoDataRequester.requestGeoShape(testActivity, prompt, waitingForDataRegistry);
+
+        assertNull(shadowActivity.getNextStartedActivity());
+        assertTrue(waitingForDataRegistry.waiting.isEmpty());
+    }
+
+    @Test
+    public void whenPermissionIsNotGranted_requestGeoTrace_doesNotLaunchAnyIntent() {
+        permissionUtils.setPermissionGranted(false);
+        activityGeoDataRequester.requestGeoTrace(testActivity, prompt, waitingForDataRegistry);
+
+        assertNull(shadowActivity.getNextStartedActivity());
+        assertTrue(waitingForDataRegistry.waiting.isEmpty());
+    }
+
+    @Test
+    public void whenPermissionIGranted_requestGeoPoint_launchCorrectIntent_andSetsFormIndexWaitingForData() {
```

Comment:
These three tests should be broken into multiple tests for the different calls to the methods under tests (for readOnly mode and different appearances for example). Otherwise, it's hard to read the test and if one fails it'll be harder to see what exactly has failed.